### DISCUSSION TITLE
Questions about Using SnowDDL in Team Workflow

### DISCUSSION THREAD

**Discussion by minghungcho** [littleK0i/SnowDDL#4] [Q&A] [answered]
When deploying with SnowDDL, it requires the a Snowflake login with SNOWDDL_ADMIN role. Although this is not an issue for CI/CD process, I am curious to learn how developers should use SnowDDL during development.

According to [Team workflow](https://docs.snowddl.com/guides/other/team-workflow?q=snowddl+admin), developers should be able to deploy to their own branch database. However, it would require granting developers to SNOWDDL_ADMIN role. What's the suggestion if this is not possible?

**Comment by littleK0i** [comment, marked as answer]
In May 2022, it is recommended to grant `SNOWDDL_ADMIN` to individual devs indeed. It might change in future, but at this moment some object types and some operations are not unavailable without `SECURITYADMIN` role.

For example:
- [SHOW USERS](https://docs.snowflake.com/en/sql-reference/sql/show-users.html) will fail;
- global roles cannot be granted to business roles;
- some types of future grants will fail;

Snowflake permission model is far from perfect in general.

In theory, if you exclude all problematic object types using `--exclude-object-types`, it may work with just a `SYSADMIN` role. But schema roles will not be created, so permission model and ownership will have to be sorted using some other way.

---

I suspect, you cannot grant `SECURITYADMIN` to individual devs because DEV and PROD environments are mixed in one Snowflake account.

A possible solution is to create a dedicated DEV Snowflake account within your organization using [CREATE ACCOUNT](https://docs.snowflake.com/en/sql-reference/sql/create-account.html). It will provide developers with "safe" sandbox environment for experiments, while PROD account remains protected from accidental harm.

**Comment by minghungcho** [comment]
In most companies, developers don't have ability to create account-level objects (warehouse, role, database), these functions usually reserved to DBA team. So development team might not want to manage those objects in config directory.

Instead, development teams has total freedom to design database objects like schemas, tables, views, functions, and etc. For that, , SnowDDL offers a superior and intuitive solution than schemachange. However, it needs a way to only allow development team to deploy (only) to their own database. So as long as SnowDDL has necessary privileges (ex as DB owner, schema owner, ..), it should be able to deploy independently.

So instead of "SnowDDL will only manage objects created by SNOWDDL_ADMIN role.", it can operate against any database as long as the --account has sufficient privilege such as database / schema owner, It think it'd a tremendous Snowflake database project tool.

**Comment by littleK0i** [comment]
Ok, I see the point.

Since SnowDDL is highly modular and implemented in pure Python, it should be easy to build a simplified execution pipeline. I can do it for you and consider including it as a standard feature afterwards.

How about the following approach?

1. A separate entry-point (e.g. `snowddl-singledb`) or plugin option (e.g. `--plugin=singledb`).
2. A separate option to set target database name.
3. Config structure remains the same, but `database_name` for all objects will be converted to target database.
4. Only schemas and schema-objects should be resolved.

Database object, ownership, future grants, roles and permissions, warehouse creation & assignment must be handled by DBA team prior to running SnowDDL in this mode.

The only permissions required for user operating SnowDDL will be:
- USAGE on DATABASE
- CREATE SCHEMA in DATABASE
- FUTURE GRANT OWNERSHIP on SCHEMA
- FUTURE GRANT OWNERSHIP on <objects_types> IN SCHEMA
- FUTURE GRANT <whatever_you_need> on <object_types> in SCHEMA
- USAGE, OPERATE on WAREHOUSE (if you want to re-create tables)

No SYSADMIN, no SECURITYADMIN.

**Reply by minghungcho** [reply]
This would be a great feature. Based on my interaction with Snowflake team, there is really no simple solution to CI/CD, let alone a database project template for new Snowflake project. Similar to a MS SQL Server Database project, I think SnowDDL provides a good foundation to any team that needs a standard project template that works with CI/CD pipeline such as Azure DevOps.
Thank you so much for such excellent project!

**Reply by littleK0i** [reply]
I have a few additional questions.

1. Do you have one "test" database object for all developers, or does every developer have its own database?
2. Is it possible to control / change names of developer databases? For example, to support the env_prefix feature: https://docs.snowddl.com/guides/other/env-prefix

I am trying to decide between two options:
- one fully customizable target database name;
- database names for individual developers are derived from production database name and uses env_prefix;

For example, if production database is called `IMPORTANT_DB`, dev databases will look like `ALICE__IMPORTAND_DB`, `BOB__IMPORTANT_DB`, etc.

Thank you.

**Reply by minghungcho** [reply]
Since we are using single Snowflake account, we do both implementations.

Scenario 1: an environment specific database. Ex. FINANACE_DEV, FINANCE_QA, FINANCE_PROD, ...
In this scenario, developers work in FINANCE_DEV, and CI/CD handles the rest of environments via SnowDDL.

Scenario 2: when developer branch from GIT, he/she will request his/her feature database. Ex. FINANACE_BOB. Once Bob completed the development, he will request a refresh of his database, FINANCE_BOB, with latest FINANCE_DEV. So Bob can validate his changes using SnowDDL. Once everything checked out, he then commit his changes to Master, and to be released to FINANCE_DEV via CI/CD.

In both scenario, databases are created beforehand via a service request. Both developers or CI/CD only focus on deploying database objects.

I believe most developers start their development by crafting changes in Snowflake WebUI or DBeaver. ex. ATLER TABLE, CREATE VIEW, .. Once they are happy with changes, they then make necessary changes in SnowDDL config folder. Then they will test by deploying changes to feature database. Once everything goods good, they proceed to do a pull request to merge the change, which triggers the CI/CD to deploy the latest commit to DEV database.

I like the idea of fully customizable target database name, which gives most flexibility to apply SnowDDL config dir to any databases.

**Reply by littleK0i** [reply]
I think I'll need a bit more time to implement this.

"Singledb" options are simple, but the actual dynamic replacement of DATABASE references is not.

It is not only about object names. Objects may have links on each other (e.g. FOREIGN KEY, STAGE, MASKING POLICY), and DATABASE must be replaced in these links as well. Such links might be hidden inside lists, dicts, another dataclasses. Links might be nested.

A smart recursive algo is required to process every property of every object, find every "database references" and replace it. I may need to rework "Identifier" helpers a little bit to make it possible to extract "DATABASE" part directly. Currently it stores just "parts" of identifier.

**Comment by littleK0i** [comment]
@minghungcho , please note, I've implemented the initial version of SingleDB mode for SnowDDL. It is available in `0.6.0`.

I'll write a full documentation page shortly. But at this moment you may use the following CLI command as a reference for initial testing:

```
snowddl-singledb --config-db=<CONFIG_DB> --target-db=<TARGET_DB> apply
```

1) It uses a separate entry-point `snowddl-singledb`.
2) `--config-db` argument is a name of specific source database in config. If only one database is present in config, it can omitted.
3) `--target-db` argument is a name of target database in Snowflake to compare and apply changes. By default it is the same as `--config-db`.

In SingleDB mode only schemas and schema objects are being resolved. No account-level objects, no roles, no grants.

In SingleDB mode ownership of `DATABASE` and `SCHEMA` objects are NOT verified. Ownership and other privileges are supposed to be managed by grants and future grants created outside of SnowDDL.

Naturally, `SYSADMIN` and `SECURITYADMIN` are not required to operate in SingleDB mode.

Please let me know if this implementation works for you.

**Comment by littleK0i** [comment]
@minghungcho , documentation for SingleDB mode is now available: https://docs.snowddl.com/single-db/overview

Enjoy!
